### PR TITLE
Store templateContent cache on disk

### PR DESCRIPTION
**Pull request by erikwilson**
Refactor catalogs to cache charts to disk, and pull chart data on
demand as needed. Files and questions are now injected at the
formatter to reduce content in CRDs, and the old templateContent
collection is deleted. Upgrade and cache cleanup methods have been
implemented to remove old data and ensure the disk does not fill up.

For /issues/14322

### CODE REVIEW

**Comment by erikwilson** [comment]
Need the following changes:
 - [ ] Add disk caching back for Git repos
 - [ ] Fetch Files & Questions on demand
 - [ ] Change traverse & update to only examine the index if possible

File: pkg/api/controllers/catalog/catalog.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	clusterCatalogController := context.Management.ClusterCatalogs("").Controller()
+	clusterCatalogController.AddHandler(ctx, "clusterCatalog", destroyClusterCatalogSync)
+
+	projectCatalogController := context.Management.ProjectCatalogs("").Controller()
+	projectCatalogController.AddHandler(ctx, "projectCatalog", destroyProjectCatalogSync)
+}
+
+func destroyCatalogSync(key string, obj *v3.Catalog) (runtime.Object, error) {
+	if obj == nil {
+		return destroy(key)
```
Comment (ibuildthecloud) [inline pkg/api/controllers/catalog/catalog.go]: There's no reason guarantee that you will see the delete of an object. What is the impact if destroy is never called?

Comment (erikwilson) [reply pkg/api/controllers/catalog/catalog.go]: If destroy is never called, and another catalog is created with the same namespace & name but different URL, then the server may analyze and use the wrong catalog cache.

Comment (erikwilson) [reply pkg/api/controllers/catalog/catalog.go]: We could verify the URL on catalog cache initialization and recreate the cache if needed. Then the only side effect of not calling destroy would be leaving some cache on disk if the catalog is deleted.

File: pkg/catalog/manager/traverse.go [hunk 2]
```diff
@@ -38,25 +41,12 @@ func (m *Manager) traverseAndUpdate(repoPath, commit string, cmt *CatalogInfo, t
...
-	for _, t := range templateContentList {
-		templateContentMap[t.Name] = struct{}{}
-	}
-
 	newHelmVersionCommits := map[string]v3.VersionCommits{}
 	var errs []error
 	var terrors []error
 	createdTemplates, updatedTemplates, deletedTemplates := 0, 0, 0
 	for chart, metadata := range index.IndexFile.Entries {
+		logrus.Infof("Catalog [%s] found chart template for [%s]", catalog.Name, chart)
```
Comment (StrongMonkey) [inline pkg/catalog/manager/traverse.go]: maybe move to debug?

File: pkg/auth/providerrefresh/refresh_daemon.go [hunk 3]
```diff
@@ -187,6 +187,11 @@ func (r *refresher) triggerUserRefresh(userName string, force bool) {
 		return
 	}
 
+	if ref == nil {
```
Comment (StrongMonkey) [inline pkg/auth/providerrefresh/refresh_daemon.go]: I don't know if this change is related to the original issue...if not we should move it to a separate pr.

Comment (erikwilson) [reply pkg/auth/providerrefresh/refresh_daemon.go]: It is not related to the original issue, but the change is needed for testing and associated with rancher/rancher/issues/17958 (@cjellick not sure if fix is sufficient to address issue). This commit can be removed before merging.

Comment (erikwilson) [reply pkg/auth/providerrefresh/refresh_daemon.go]: Fix was already merged, will remove & rebase

File: pkg/catalog/helm/setup.go [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+		return nil, errors.New("Catalog is not defined for helm")
+	}
+
+	h := newCache(namespace, catalog.Name)
+	h.url = catalog.Spec.URL
+	h.branch = catalog.Spec.Branch
+	h.lastCommit = catalog.Status.Commit
+	h.username = catalog.Spec.Username
+	h.password = catalog.Spec.Password
+	h.Lock()
```
Comment (StrongMonkey) [inline pkg/catalog/helm/setup.go]: Is it better to move the lock into specific read and write of cache files? i feel a little bit lost on where the lock is acquired and released...

Comment (erikwilson) [reply pkg/catalog/helm/setup.go]: I agree the locking stuff can use some rework. However, whenever we create a New helm object we can assume there will be some I/O involved and we could Lock by default and Unlock when done. When using scaledContext to destroy, the lock does not seem to be effective with traverseAndUpdate (when modified appropriately), so may want to move to file locking instead.

Comment (erikwilson) [reply pkg/catalog/helm/setup.go]: Actually, traverseAndUpdate is probably a special case where we need to check if the cache has been destroyed after re-locking, since the process is long-lived.

File: pkg/controllers/management/catalog/controller.go [hunk 5]
```diff
@@ -16,7 +15,7 @@ import (
 
 func Register(ctx context.Context, management *config.ManagementContext) {
```
Comment (StrongMonkey) [inline pkg/controllers/management/catalog/controller.go]: Looks like this is still running in master controller, which is only running in a single leader(so cache will only exist inside leader instance) @erikwilson have you tested git catalog on a HA setup?

Comment (StrongMonkey) [reply pkg/controllers/management/catalog/controller.go]: wait... I get it now.

**Comment by StrongMonkey** [comment]
TemplateContent is also used by helm controller to store revisions of catalog app.(https://github.com/rancher/rancher/blob/a88b088e843c08c91f990021cd0d4cf283386563/pkg/controllers/user/helm/controller.go#L284). So when you upgrade and try to clean up templateContent, make sure this part don't break.

File: pkg/api/customization/catalog/templateversion.go [hunk 6]
```diff
@@ -71,6 +61,43 @@ func (t TemplateVerionFormatterWrapper) TemplateVersionFormatter(apiContext *typ
 		delete(resource.Values, "upgradeVersionLinks")
 		resource.Values["upgradeVersionLinks"] = linkMap
 	}
+
+	templateVersion := &client.CatalogTemplateVersion{}
```
Comment (StrongMonkey) [inline pkg/api/customization/catalog/templateversion.go]: can you unmarshall resource.Values to templateVersion? you can save one api call.

File: pkg/controllers/user/helm/common.go [hunk 7]
```diff
@@ -118,36 +137,36 @@ func generateTemplates(obj *v3.App, templateVersionClient mgmtv3.CatalogTemplate
...
+	commands := append([]string{"template", appDir, "--name", obj.Name, "--namespace", obj.Spec.TargetNamespace}, setValues...)
 
 	cmd := exec.Command(helmName, commands...)
 	sbOut := &bytes.Buffer{}
 	sbErr := &bytes.Buffer{}
 	cmd.Stdout = sbOut
 	cmd.Stderr = sbErr
 	if err := cmd.Start(); err != nil {
-		return "", "", "", errors.Wrapf(err, "helm template failed. %s", filterErrorMessage(sbErr.String(), dir, "template-dir"))
+		return "", "", "", tempDir, errors.Wrapf(err, "helm template failed. %s", sbErr.String())
```
Comment (StrongMonkey) [inline pkg/controllers/user/helm/common.go]: Did you intentionally remove filterErrorMessage? It was added to fix another problem to avoid random message being written into etcd.

File: pkg/controllers/user/helm/controller.go [hunk 8]
```diff
@@ -166,12 +174,14 @@ func (l *Lifecycle) DeployApp(obj *v3.App) (*v3.App, error) {
 		}
 	}
 	newObj, err := v3.AppConditionInstalled.Do(obj, func() (runtime.Object, error) {
-		template, notes, tempDir, err := generateTemplates(obj, l.TemplateVersionClient, l.TemplateContentClient)
-		defer os.RemoveAll(tempDir)
+		template, notes, appDir, tempDir, err := l.generateTemplates(obj)
+		if tempDir != "" {
```
Comment (StrongMonkey) [inline pkg/controllers/user/helm/controller.go]: Shouldn't really matter if dir is empty or not.

**Comment by erikwilson** [comment]
Some metrics on the impact of this change from local testing:

### Master
Library catalog initial setup: ~20s
Helm catalog setup: 13m initial, 4m second update, 4s third update = ~17m
Size of disk cache: 18MB
Size of kubernetes resource dump as json to disk: 80MB (2.7MB catalogtemplates, 22MB catalogtemplateversions, 53MB templatecontents)

### PR branch
Library catalog initial setup: ~15s
Helm catalog initial setup: ~1m
Size of disk cache: 18MB
Size of kubernetes resource dump as json to disk: 12MB (3MB catalogtemplates, 6.6MB catalogtemplateversions, 0MB templatecontents)
Number of template contents deleted: 22654 (~2m)

File: pkg/api/controllers/catalog/catalog.go (hunk 1 above)
Comment (erikwilson) [reply pkg/api/controllers/catalog/catalog.go]: This PR has been updated to scan through the cache directory and remove any directories which do not correspond with a catalog, it will clean up stale items in the cache once at start, on a period of every hour, and catalog changes are debounced to clean up after a minute of inactivity.

File: pkg/catalog/manager/traverse.go (hunk 2 above)
Comment (erikwilson) [reply pkg/catalog/manager/traverse.go]: Fixed and associated logging for this PR cleaned up

File: pkg/catalog/helm/setup.go (hunk 4 above)
Comment (erikwilson) [reply pkg/catalog/helm/setup.go]: Locking stuff has been revamped

File: pkg/api/customization/catalog/templateversion.go (hunk 6 above)
Comment (erikwilson) [reply pkg/api/customization/catalog/templateversion.go]: Unmarshalled!

File: pkg/controllers/user/helm/common.go (hunk 7 above)
Comment (erikwilson) [reply pkg/controllers/user/helm/common.go]: Was an accident, thanks for finding that!

File: pkg/controllers/user/helm/controller.go (hunk 8 above)
Comment (erikwilson) [reply pkg/controllers/user/helm/controller.go]: Wasn't sure of the behavior, but tested to verify that nothing is removed for empty string, thanks for pointing this out!

**Comment by erikwilson** [comment]
This is probably ready for some more review @StrongMonkey & @cjellick, especially re cache cleanup and removing old template contents.

File: pkg/catalog/manager/manager.go [hunk 9]
```diff
@@ -53,3 +59,40 @@ func (m *Manager) deleteChart(toDelete string, namespace string) error {
 	}
 	return nil
 }
+
+func (m *Manager) DeleteOldTemplateContent() {
+	templateContents, err := m.templateContentClient.List(metav1.ListOptions{})
+	if err != nil {
+		logrus.Error(err)
```
Comment (StrongMonkey) [inline pkg/catalog/manager/manager.go]: What happens if there is an error? will the code never be executed until next restart? Should we retry this if error happens?

File: pkg/catalog/manager/traverse.go [hunk 10]
```diff
@@ -199,6 +197,11 @@ func (m *Manager) traverseAndUpdate(helm *helmlib.Helm, commit string, cmt *Cata
 		if catalog == nil || catalog.Name == "" {
 			return errors.New("Catalog is no longer available")
 		}
+		if commit == catalog.Status.Commit && v3.CatalogConditionUpgraded.IsTrue(catalog) {
```
Comment (StrongMonkey) [inline pkg/catalog/manager/traverse.go]: I think this is already done in https://github.com/rancher/rancher/blob/f8a2e3830b3a80afc9cd0f3691168a043492f8f1/pkg/catalog/manager/catalog_sync.go#L42-L52 to avoid update if commit is the same. Is there anything special to add this check?

File: pkg/catalog/manager/manager.go (hunk 9 above)
Comment (erikwilson) [reply pkg/catalog/manager/manager.go]: The code won't be called again until next restart, but afaik it shouldn't fail unless the k8s api server is unavailable. My bigger concern is for the memory & time consumption of this method.

File: pkg/catalog/manager/traverse.go (hunk 10 above)
Comment (erikwilson) [reply pkg/catalog/manager/traverse.go]: We reload the catalog in traverseAndUpdate because of the time required to perform the operation, and it appears sync may have queued up events before we have successfully marked the catalog as processed. Since traverseAndUpdate may have stale catalog data we should verify the catalog has not been processed after refreshing to end updating as soon as possible.

File: pkg/catalog/helm/helm.go [hunk 11]
```diff
@@ -174,65 +214,173 @@ func FetchTgz(url, username, password string) ([]v3.File, error) {
...
-			newFile, err := LoadFile(version, strings.TrimPrefix(url, "file://"))
-			if err != nil {
-				return nil, err
-			}
-			files = append(files, *newFile)
-			continue
+	for _, file := range version.LocalFiles {
+		newFile, err := h.loadFile(version, file)
+		if err != nil {
+			logrus.Errorf("Error loading helm file! [%s]: %s", file, err)
```
Comment (StrongMonkey) [inline pkg/catalog/helm/helm.go]: I think in general errors should be returned and handled on the top level, instead of printing in here.

File: pkg/catalog/helm/helm.go [hunk 12]
```diff
@@ -174,65 +214,173 @@ func FetchTgz(url, username, password string) ([]v3.File, error) {
...
-		newFiles, err := FetchTgz(url, username, password)
+		content, err := ioutil.ReadFile(filename)
 		if err != nil {
+			return err
+		}
+		filemap[name] = string(content)
+		return nil
+	})
+	if err != nil {
+		logrus.Errorf("Failed to load chart files: %s", err)
```
Comment (StrongMonkey) [inline pkg/catalog/helm/helm.go]: same above

File: pkg/catalog/helm/helm.go [hunk 13]
```diff
@@ -174,65 +214,173 @@ func FetchTgz(url, username, password string) ([]v3.File, error) {
...
+	return false
+}
+
+func (h *Helm) fetchURLs(urls []string) ([]v3.File, error) {
+	for _, url := range urls {
+		newFiles, err := h.fetchTgz(url)
+		if err == nil {
+			return newFiles, nil
+		}
+		logrus.Errorf("Error fetching helm url [%s]: %s", url, err)
```
Comment (StrongMonkey) [inline pkg/catalog/helm/helm.go]: Specific error not being returned. You can use errors.Wrapf to wrap error.

File: pkg/catalog/manager/traverse.go [hunk 14]
```diff
@@ -231,15 +225,16 @@ func (m *Manager) traverseAndUpdate(repoPath, commit string, cmt *CatalogInfo, t
 	}
 	// delete non-existing templates
 	for _, toDelete := range toDeleteChart {
-		logrus.Debugf("Deleting template %s and its associated templateVersion", toDelete)
+		logrus.Debugf("Deleting template %s and its associated templateVersion in namespace %s", toDelete, templateNamespace)
 		if err := m.deleteChart(toDelete, templateNamespace); err != nil {
-			return err
+			logrus.Error(err)
```
Comment (StrongMonkey) [inline pkg/catalog/manager/traverse.go]: is there a reason to not return this error?

File: pkg/controllers/management/catalog/controller.go [hunk 15]
```diff
@@ -79,6 +80,9 @@ func Run(ctx context.Context, cacheRoot string, refreshInterval int, management
 	clusterCatalogController := management.Management.ClusterCatalogs("").Controller()
 	clusterCatalogController.AddHandler(ctx, "clusterCatalog", m.ClusterCatalogSync)
 
+	go m.DeleteOldTemplateContent()
+	go m.DeleteBadCatalogTemplates()
```
Comment (StrongMonkey) [inline pkg/controllers/management/catalog/controller.go]: need to address the scenario where error happens and the code will never be executed until server restarts

File: pkg/catalog/manager/manager.go [hunk 16]
```diff
@@ -1,236 +1,204 @@
...
-			if err = git.Update(repoPath, branch); err != nil {
-				return "", "", errors.Wrap(err, "Update failed")
+	}
+
+	var deleteCount int
+	for _, templateContent := range templateContents.Items {
+		digest := templateContent.Name
+		if !appTags[digest] {
+			if err := m.templateContentClient.Delete(digest, &metav1.DeleteOptions{}); err != nil {
+				logrus.Error(err)
```
Comment (StrongMonkey) [inline pkg/catalog/manager/manager.go]: errors should be aggregated and returned

File: pkg/catalog/manager/manager.go [hunk 17]
```diff
@@ -1,236 +1,204 @@
...
+			catalogName = template.Spec.ProjectCatalogID
+		}
+
+		ns, name := helmlib.SplitNamespaceAndName(catalogName)
+		if ns == "" {
+			ns = namespace.GlobalNamespace
+		}
+		if !hasCatalog[getKey(ns, name)] {
+			if err := m.templateClient.DeleteNamespaced(template.Namespace, template.Name, &metav1.DeleteOptions{}); err != nil {
+				logrus.Error(err)
```
Comment (StrongMonkey) [inline pkg/catalog/manager/manager.go]: same above

File: pkg/catalog/manager/manager.go [hunk 18]
```diff
@@ -1,236 +1,204 @@
...
+		if err != nil {
+			logrus.Errorf("Catalog-manager error extracting namespace/name from template version: %s", err)
+			continue
+		}
+		if ns == "" {
+			ns = namespace.GlobalNamespace
+		}
+		if !hasCatalog[getKey(ns, name)] {
+			if err := m.templateVersionClient.DeleteNamespaced(templateVersion.Namespace, templateVersion.Name, &metav1.DeleteOptions{}); err != nil {
+				logrus.Error(err)
```
Comment (StrongMonkey) [inline pkg/catalog/manager/manager.go]: same above

File: pkg/catalog/manager/manager.go (hunk 9 above)
Comment (StrongMonkey) [reply pkg/catalog/manager/manager.go]: But we should still retry this method until it succeeds. There are plenty of reasons k8s api call could fail. Otherwise the cache will never be cleaned up.

File: pkg/catalog/manager/traverse.go (hunk 14 above)
Comment (erikwilson) [reply pkg/catalog/manager/traverse.go]: Good catch! 

I think there might be a bug in norman, in combination with the traverse code in master where returning an error here causes an endless loop of unrecoverable errors. In master code when we refresh the catalog we don't check if the catalog has been removed, so disabling a catalog while it is updating may cause some weird values to be used. 

I think if we attempt to delete a namespaced resource without a namespace the resource became updated but not removed, and then future attempts to remove it would cause the continuous errors. I didn't dig super far into the conditions or root cause of what triggered the loop.

**Comment by StrongMonkey** [comment]
LGTM assuming build is green.

**Comment by cjellick** [comment]
LGTM
